k2/csrc/context.cu

```diff
@@ -28,4 +28,47 @@ RegionPtr NewRegion(ContextPtr &context, std::size_t num_bytes) {
   return ans;
 }
 
+ParallelRunner::ParallelRunner(ContextPtr c) : c_(c) {
+  if (c_->GetDeviceType() == kCuda) {
+    auto ret = cudaEventCreate(&event_);
+    K2_CHECK_CUDA_ERROR(ret);
+    // record event on `c_->GetCudaStream` and will be waited on `NewStream``
+    ret = cudaEventRecord(event_, c_->GetCudaStream());
+    K2_CHECK_CUDA_ERROR(ret);
+  }
+}
+cudaStream_t ParallelRunner::NewStream() {
+  DeviceType d = c_->GetDeviceType();
+  if (d == kCpu) {
+    return kCudaStreamInvalid;
+  } else {
+    K2_CHECK_EQ(d, kCuda);
+    cudaStream_t stream;
+    auto ret = cudaStreamCreate(&stream);
+    K2_CHECK_CUDA_ERROR(ret);
+    streams_.push_back(stream);
+
+    ret = cudaStreamWaitEvent(stream, event_, 0);
+    K2_CHECK_CUDA_ERROR(ret);
+    return stream;
+  }
+}
+
+ParallelRunner::~ParallelRunner() {
+  if (c_->GetDeviceType() == kCuda) {
+    for (std::size_t i = 0; i != streams_.size(); ++i) {
+      // record event on `stream_[i]` and wait on c_->GetCudaStream
+      auto ret = cudaEventRecord(event_, streams_[i]);
+      K2_CHECK_CUDA_ERROR(ret);
+      ret = cudaStreamWaitEvent(c_->GetCudaStream(), event_, 0);
+      K2_CHECK_CUDA_ERROR(ret);
+      ret = cudaStreamDestroy(streams_[i]);
+      K2_CHECK_CUDA_ERROR(ret);
```

Comment:
I think we need to destroy these events too, inside the loop.
I had a look at the documentation, it doesn't matter that they are not finished with yet.

Reply:
There is only one event_ now and I destroy it at the end of loop. Do you mean we should create and destroy the `event_` again and again in the loop？

Reply:
Fixed.